### PR TITLE
Added Partner Info to Submission Download

### PR DESCRIPTION
**Pull request by alvinwan**
@Sumukh @soumyabasu Ready.

@jeffreylu017 Tada. (see gif below). Students without a group just won't have a `group.json`

get member info

### CODE REVIEW

File: server/app/utils.py [hunk 1]
```diff
@@ -517,12 +517,31 @@ def add_subm_to_zip(subm, Submission, zipfile, result):
...
+    file_contents[file_name] = file_content
+
+# TODO(Alvin): generalize, cleanup everything about zip
+def backup_group_file(backup):
+    """ Returns group information: group_[group ID], group JSON """
+    G = ModelProxy.Group
+    group = G.lookup(backup.submitter, backup.assignment)
+    if group:
+        json_data = group.to_json()
+        order = {chr(97+i): u['email'][0]
```
Comment (Sumukh) [inline server/app/utils.py]: might want to leave a comment that the chr(97+idx) is to get A/B instead of 0/1.

Comment (alvinwan) [inline server/app/utils.py]: Sure thing. Adding.

**Comment by Sumukh** [comment]
What's the behavior on a bulk download? Going to be a group_members.json file in ever group on a bulk download?  

We might want to just standardize this into something like a meta.json that includes other information like who submitted it, what time it was submitted, partners in the group etc. If that's not much more work - I think it's worth it. 

It might also be useful to have this is in the exported grades. (but that can wait for later)

File: server/app/utils.py [hunk 2]
```diff
@@ -517,12 +517,33 @@ def add_subm_to_zip(subm, Submission, zipfile, result):
...
+        # 97+i converts 0, 1, 2, 3 to ascii codes corresponding to a, b, c...
+        # chr converts ascii code to an ascii char
+        order = {chr(97+i): u['email'][0]
+                 for i, u in enumerate(json_data['member'])}
+        return 'group_members_%s.json' % group.key.id(), str(json.dumps(order))
+
+
 def make_zip_filename(user, now):
     """ Makes zip filename: query_USER EMAIL_DATETIME.zip """
     outlawed = [' ', '.', ':', '@']
```
Comment (Sumukh) [inline server/app/utils.py]: '/' should also be outlawed right?

**Comment by Sumukh** [comment]
@jeffreylu017 - was this what you had in mind/would work for you?

**Comment by alvinwan** [comment]
`group_members.json` is present in any submission folder that has a group.

I originally did that, actually. I just spit out the entire JSON, but it's incredibly ugly and unreadable... if you say so though, it's an easy add. I'll add in a `group_meta.json` and a `submission_meta.json`.

**Comment by Sumukh** [comment]
json.dump has the following optional arguments which make it a little prettier and easier to read `sort_keys=True, indent=4, separators=(',', ': ')`

**Comment by alvinwan** [comment]
Oh, thanks. Shoulda read the docs.

**Comment by soumyabasu** [comment]
I'm not 100% sure if we should convert to Partner A and Partner B since that's a 61A naming convention. We should just give back the group ordering in a nice way (I think 0/1 is fine actually) and the course staff should make the conversion into whatever they want.

**Comment by alvinwan** [comment]
@soumyabasu Mhm. Fair enough; any group with more than 26 members would start to have gibberish for the keys in `group_members.json` anyhow.

File: server/app/utils.py (hunk 2 above)
Comment (alvinwan) [inline server/app/utils.py]: Oh, yes. Oops.

That'd only happen if someone had a `/` in his/her email address though. Added it anyhow.

**Comment by alvinwan** [comment]
Final export results. @soumyabasu @Sumukh Are we good to go?

metas

**Comment by jeffreylu017** [comment]
Sorry I'm late to the party.

Will this work if I export this using the OK client (`python3 ok --export --latest`)?  I'm unable to unzip the file downloaded from the website, so I've been downloading submissions through the terminal.

The json file looks great.  I'm flexible with any format, and I can just grep the information I need.

**Comment by alvinwan** [comment]
Let me check; not sure what endpoint the client uses.

**Comment by alvinwan** [comment]
@jeffreylu017 For download, you'll have to use Safari for now. Not sure why it downloads differently, sorry.

`python3 ok --export --latest` doesn't call the search download endpoint, so it won't include these meta files.

**Comment by alvinwan** [comment]
@albert12132 @jeffreylu017 Once sumukh/soumya approve and merge in the two PRs I made, zip downloads will work on all browsers... this entire time, I forgot to `zipfile.close()`.

**Pull request merged**